porter-dev/porter
# SyncedEnvGroup Revamp

Environment groups are stored in `porter-env-group` namespace with a the name `environment-group-name.environment-group-version` and a corresponding label `porter.run/environment-group-name` and `porter.run/environment-group-version` which are used for querying.

An envgroup consists of a configmap which contains all non-secret values, and a secret which has all secret values. If an env group only has non-secret values, a secret will be created anyway for parity, and vice-versa.

Applications are linked to preview groups by adding a label called `porter.run/linked-environment-group`. This uses a dot separator which refers to the non-versioned environment-group-name. 
At deploy-time, we check the label on the applications, and ensure that the `porter-env-group` environment group parts are synced into the target environment. These can then be referenced as a configMapRef and a secretRef. We chose this pattern as we can use a CSI in future to move the secrets and configs to a 3rd party provider of the customer's choice. This is not on the roadmap, but just given for context.

To list all applications that are linked to an application, we just search all deployments and cronjobs for the label `porter.run/linked-environment-group` which should have a value which equals a value in the `porter-env-group` namespace with the label `porter.run/environment-group-name`.

To override an environment variable in an application which has an env group attached, we just inject this into the k8s manifest as a raw value. 

**Testing**:
Every step below should end in the following steps:
1. Ensure the label is in the helm values, and the configMapRef and secretMapRef exists
2. Ensure pod spins up and contains all values from the env group, and expected overrides

Scenarios:
------
1. Create an env group
2. Create an application and attach the env groups to a web chart
3. Create an application and attach the env groups to a job chart
4. Create an application and attach the env groups to a worker chart
------
1. Create another env group
2. Create another application attaching both env groups to it
------

1. Create another application with a web, worker, and job service
2. Attach both env groups to this service
------

1. Create another application with a web, worker, and job service
2. Attach both env groups to this service
3. Add an override environment variable to the application
------

1. Push an update to the environment group which was not overridden
2. Ensure that all above applications were updated and that the override stays
------

1. Push an update to the environment group which was overridden
2. Ensure that all above applications were updated and that the override stays
------

1. Create an application with only a job
2. Attach an env group
3. Add a new service to the application
